lib/client/session.go

```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+			if err := cmd.GenerateUntrustedCookie(ns.x11Display, ns.x11Proto, ns.nodeClient.TC.X11ForwardingTimeout); err != nil {
+				return trace.Wrap(err)
+			}
+
+			xauthEntry, err = x11.NewXAuthCommand(ctx, xauthFile.Name()).ReadEntry(ns.x11Display)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	ns.x11Proto, ns.x11RealCookie = xauthEntry.Proto, xauthEntry.Cookie
+
+	// The client's xauth cookie should never be sent to the server,
+	// so instead we create a spoof of the cookie to authenticate the server.
+	// During forwarding, the fake cookie will be replaced with the real
+	// cookie in order to connect to the client's local X display.
+	ns.x11FakeCookie, err = xauthEntry.SpoofCookie()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// serveX11Channels serves incoming x11 channels by starting x11 forwarding with the session.
+func (ns *NodeSession) serveX11Channels(ctx context.Context, sess *ssh.Session) error {
+	err := x11.ServeX11ChannelRequests(ctx, ns.nodeClient.Client, func(ctx context.Context, nch ssh.NewChannel) {
+		if !ns.x11RefuseTime.IsZero() && time.Now().After(ns.x11RefuseTime) {
+			nch.Reject(ssh.Prohibited, "rejected x11 channel request after ForwardX11Timeout")
+		}
+
+		var req x11.X11ChannelRequestPayload
+		if err := ssh.Unmarshal(nch.ExtraData(), &req); err != nil {
+			nch.Reject(ssh.Prohibited, "invalid payload")
+			log.Debug("rejected x11 channel request with invalid payload")
+			return
+		}
+
+		log.Debugf("received x11 channel request from %s:%d", req.OriginatorAddress, req.OriginatorPort)
+		xchan, sin, err := nch.Accept()
+		if err != nil {
+			log.WithError(err).Debug("failed to accept x11 channel request")
+			return
+		}
+		defer xchan.Close()
+
+		// The x11 channel's initial auth packet is scanned for the fakeCookie.
+		// If the cookie is present, it will be replaced with the real cookie.
+		// Otherwise, an access denied error will be returned.
+		authPacket, err := x11.ReadAndRewriteXAuthPacket(xchan, ns.x11Proto, ns.x11FakeCookie, ns.x11RealCookie)
+		if trace.IsAccessDenied(err) {
+			log.Error("x11 connection rejected due to wrong authentication")
+			return
+		} else if err != nil {
+			log.WithError(err).Debug("Failed to read auth packet from x11 channel request")
+			return
+		}
+
+		xconn, err := x11.DialXDisplay(ns.x11Display)
+		if err != nil {
+			log.WithError(err).Debug("Failed to connect to client's display")
+			return
+		}
+
+		// send the connection the pre-processed x11 auth packet
+		if _, err := xconn.Write(authPacket); err != nil {
+			log.WithError(err).Debug("Failed to write xauth packet")
+			return
+		}
+
+		// Forward ssh requests on the x11 channel until x11 forwarding is complete
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		go func() {
+			if err := sshutils.ForwardRequests(ctx, sin, sess); err != nil {
+				log.WithError(err).Debug("Failed to forward ssh request during x11 forwarding")
```

Comment:
Should you check the error here and not log if it is `context.Canceled`?

Reply:
if the context is canceled this should return nil for error

Reply:
Adding this check inside `ForwardRequests` instead